### PR TITLE
#184347040 Searching for a product

### PR DESCRIPTION
**Pull request by nshimiyejayd**
#### What does this PR do?
- This PR implements the needs of user for searching a product

#### Description of Task to be completed?
- User can search a product by name ,by category, by( min or max) price and by expire date criteria
- User can search a product by combination of criteria

#### How should this be manually tested?
- Clone [repo](https://github.com/atlp-rwanda/vikings-ec-bn)
- Checkout branch `feature/search-product-184347040`
- Start server
- Open `http://localhost:5000/api-docs` in your browser
- Try `/api/v1/products/search`

#### What are the relevant pivotal tracker stories?
[#184347040](https://www.pivotaltracker.com/story/show/184347040)

#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

### CODE REVIEW

File: src/middlewares/product.middleware.js [hunk 1]
```diff
@@ -41,3 +43,33 @@ export const checkNumberOfImages = (req, res, next) => {
     return res.status(400).json({ message: 'Please insert at least 4 images' });
   next();
 };
+export const searchCriteriasFormat = async (req, res, next) => {
+  const hasSearchParams = Object.values(req.query).some(val => val !== undefined && val !== '');
```
Comment (irakozetony) [inline src/middlewares/product.middleware.js]: Take this to validation

**Review by irakozetony — CHANGES_REQUESTED** [review]

**Review by ndatumuremyi — APPROVED** [review]

**Review by irakozetony — APPROVED** [review]

**Review by fabmukunzi — APPROVED** [review]

File: src/controllers/product.controller.js [hunk 2]
```diff
@@ -21,3 +21,18 @@ export const createProduct = async (req, res) => {
       });
   }
 };
+
+export const searchProductController = async(req, res) => {
+  const searchQuery = req.query;
+  try {
+   const result = await ProductService.searchProduct(searchQuery);
+   return res.status(200).json(result);
```
Comment (TuyizeeAnastase) [inline src/controllers/product.controller.js]: this is not how you send json

File: src/services/product.service.js [hunk 3]
```diff
@@ -1,7 +1,13 @@
...
 
 export class ProductService {
   static async createProduct(product) {
     return await Products.create(product);
   }
+  static async searchProduct(searchQuery){
+    return await Products.findAll({
+      where: searchQuery,
+      include: [{ model: Categories }]
+    });
```
Comment (TuyizeeAnastase) [inline src/services/product.service.js:47]: use sequelize like Operation not where.  [Model querying](https://sequelize.org/docs/v6/core-concepts/model-querying-finders/)

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

**Review by Abdden — APPROVED** [review]

File: src/services/product.service.js (hunk 3 above)
Comment (nshimiyejayd) [reply src/services/product.service.js:47]: Here I have used `like operator` in middleware where I formatted that object `searchQuery`

File: src/utils/searchCategoryByName.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { Categories } from '../database/models/index';
+
+export const isSearchCategoryExists = async(categoryName) => {
+    return await Categories.findOne({
+      where: { name: categoryName.toLowerCase() },
+    });
+};
```
Comment (ncutixavier) [inline src/utils/searchCategoryByName.js]: This should be in the service

**Review by ncutixavier — CHANGES_REQUESTED** [review]

File: src/utils/searchCategoryByName.js (hunk 4 above)
Comment (ncutixavier) [inline src/utils/searchCategoryByName.js]: Why didn't you remove this file?

**Review by ncutixavier — CHANGES_REQUESTED** [review]

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**